File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/auth/model/Service.java
package uk.gov.justice.digital.hmpps.oauth2server.auth.model;

import com.google.common.base.Splitter;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import org.apache.commons.lang3.StringUtils;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.Table;
import java.util.Collections;
import java.util.List;

@Entity
@Table(name = "OAUTH_SERVICE")
@Data
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(of = {"code"})
public class Service {
    @Id
    @Column(nullable = false)
    private String code;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private String description;

    @Column(name = "authorised_roles")
    private String authorisedRoles;

    @Column(nullable = false)
    private String url;

    @Column(nullable = false)
    private boolean enabled;

    @Column
    private String email;

    public List<String> getRoles() {
        //noinspection UnstableApiUsage
        return StringUtils.isBlank(authorisedRoles) ? Collections.emptyList() : Splitter.on(',').trimResults().splitToList(authorisedRoles);
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/model/Service.kt
package uk.gov.justice.digital.hmpps.oauth2server.auth.model

import org.apache.commons.lang3.StringUtils.trimToNull
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.Table

@Entity
@Table(name = "OAUTH_SERVICE")
data class Service(
  @Id
  @Column(nullable = false)
  val code: String,
  @Column(nullable = false)
  val name: String,

  @Column(nullable = false)
  val description: String,

  @Column(name = "authorised_roles")
  private var authorisedRoles: String? = null,

  @Column(nullable = false)
  val url: String,

  @Column(nullable = false)
  val enabled: Boolean = false,

  @Column
  val email: String? = null,
) {

  val roles: List<String>
    get() = authorisedRoles?.split(',')?.mapNotNull { trimToNull(it) } ?: emptyList()

  val isUrlInsteadOfEmail: Boolean
    get() = email?.startsWith("http") ?: false

  var authorisedRolesWithNewlines: String
    get() = authorisedRoles?.replace(",".toRegex(), "\n") ?: ""
    set(authorisedRolesWithNewlines) {
      authorisedRoles = authorisedRolesWithNewlines
        .replace("\n".toRegex(), ",")
        .split(',')
        .mapNotNull { trimToNull(it) }
        .map { it.uppercase() }
        .map { if (it.startsWith("ROLE_")) it else "ROLE_$it" }
        .joinToString(",")
    }
}
